File: internal/secretbox/authcache.go
package secretbox

import (
	"encoding/binary"
	"time"
)

const DefaultCacheDuration int64 = 600

// AuthCache provides per-app, per-secret authentication caching with HMAC integrity.
type AuthCache struct {
	db              *SecretDatabase
	authKeyProvider func() ([]byte, error)
	cacheDuration   int64
	TimeProvider    func() int64
}

type CacheOption func(*AuthCache)

func WithCacheDuration(d int64) CacheOption {
	return func(c *AuthCache) { c.cacheDuration = d }
}

func WithTimeProvider(tp func() int64) CacheOption {
	return func(c *AuthCache) { c.TimeProvider = tp }
}

func NewAuthCache(db *SecretDatabase, authKeyProvider func() ([]byte, error), opts ...CacheOption) *AuthCache {
	c := &AuthCache{
		db:              db,
		authKeyProvider: authKeyProvider,
		cacheDuration:   DefaultCacheDuration,
		TimeProvider:    func() int64 { return time.Now().Unix() },
	}
	for _, opt := range opts {
		opt(c)
	}
	return c
}

func (c *AuthCache) IsValid(callerID, secretName string) bool {
	key, err := c.authKeyProvider()
	if err != nil {
		return false
	}

	conn, err := c.db.Connection()
	if err != nil {
		return false
	}

	var timestamp int64
	var storedHMAC []byte
	err = conn.QueryRow(
		`SELECT timestamp, hmac FROM auth_cache WHERE secret_name = ? AND caller_id = ?`,
		secretName, callerID,
	).Scan(&timestamp, &storedHMAC)
	if err != nil {
		return false
	}

	now := c.TimeProvider()
	if now-timestamp > c.cacheDuration || timestamp > now {
		return false
	}

	timestampData := make([]byte, 8)
	binary.LittleEndian.PutUint64(timestampData, uint64(timestamp))

	message := BuildMessage(timestampData, callerID, secretName)
	expectedHMAC := HMACSHA256(key, message)

	return ConstantTimeEqual(expectedHMAC, storedHMAC)
}

func (c *AuthCache) Update(callerID, secretName string) {
	key, err := c.authKeyProvider()
	if err != nil {
		return
	}

	conn, err := c.db.Connection()
	if err != nil {
		return
	}

	timestamp := c.TimeProvider()
	timestampData := make([]byte, 8)
	binary.LittleEndian.PutUint64(timestampData, uint64(timestamp))

	message := BuildMessage(timestampData, callerID, secretName)
	hmacValue := HMACSHA256(key, message)

	_, _ = conn.Exec(
		`INSERT OR REPLACE INTO auth_cache (secret_name, caller_id, timestamp, hmac) VALUES (?, ?, ?, ?)`,
		secretName, callerID, timestamp, hmacValue,
	)
}

func (c *AuthCache) Invalidate(secretName string) {
	conn, err := c.db.Connection()
	if err != nil {
		return
	}
	_, _ = conn.Exec(`DELETE FROM auth_cache WHERE secret_name = ?`, secretName)
}

// BuildMessage constructs the HMAC message for cache validation.
// Format: timestamp (8 bytes LE) || caller_length (4 bytes BE) || caller_id || secret_name
func BuildMessage(timestampData []byte, callerID, secretName string) []byte {
	callerData := []byte(callerID)
	nameData := []byte(secretName)

	callerLen := make([]byte, 4)
	binary.BigEndian.PutUint32(callerLen, uint32(len(callerData)))

	message := make([]byte, 0, len(timestampData)+4+len(callerData)+len(nameData))
	message = append(message, timestampData...)
	message = append(message, callerLen...)
	message = append(message, callerData...)
	message = append(message, nameData...)

	return message
}

File: Sources/SecretBoxLib/AuthCache.swift
import CryptoKit
import Foundation
import SQLite

public class AuthCache {
    public static let defaultCacheDuration: Int64 = 600

    private let db: SecretDatabase
    private let authKeyProvider: () throws -> SymmetricKey
    private let cacheDuration: Int64
    let timeProvider: () -> Int64

    // SQLite column definitions
    private let cacheTable = Table("auth_cache")
    private let secretNameCol = SQLite.Expression<String>("secret_name")
    private let callerIdCol = SQLite.Expression<String>("caller_id")
    private let timestampCol = SQLite.Expression<Int64>("timestamp")
    private let hmacCol = SQLite.Expression<Data>("hmac")

    public init(
        db: SecretDatabase,
        authKeyProvider: @escaping () throws -> SymmetricKey,
        cacheDuration: Int64 = defaultCacheDuration,
        timeProvider: @escaping () -> Int64 = { Int64(time(nil)) }
    ) {
        self.db = db
        self.authKeyProvider = authKeyProvider
        self.cacheDuration = cacheDuration
        self.timeProvider = timeProvider
    }

    public func isValid(for callerID: String, secretName: String) -> Bool {
        guard let key = try? authKeyProvider(),
              let conn = try? db.connection else { return false }

        guard let row = try? conn.pluck(
            cacheTable.filter(secretNameCol == secretName && callerIdCol == callerID)
        ) else { return false }

        let timestamp = row[timestampCol]
        let storedHmac = row[hmacCol]

        let now = timeProvider()
        guard now - timestamp <= cacheDuration, timestamp <= now else { return false }

        var ts = timestamp.littleEndian
        let timestampData = Data(bytes: &ts, count: MemoryLayout<Int64>.size)
        let message = Self.buildMessage(timestampData: timestampData, callerID: callerID, secretName: secretName)
        let expectedHmac = hmacSHA256(key: key, data: message)

        return constantTimeEqual(expectedHmac, storedHmac)
    }

    public func update(for callerID: String, secretName: String) {
        guard let key = try? authKeyProvider(),
              let conn = try? db.connection else { return }

        let timestamp = timeProvider()
        var ts = timestamp.littleEndian
        let timestampData = Data(bytes: &ts, count: MemoryLayout<Int64>.size)

        let message = Self.buildMessage(timestampData: timestampData, callerID: callerID, secretName: secretName)
        let hmac = hmacSHA256(key: key, data: message)

        _ = try? conn.run(cacheTable.insert(or: .replace,
            secretNameCol <- secretName,
            callerIdCol <- callerID,
            timestampCol <- timestamp,
            hmacCol <- hmac
        ))
    }

    public func invalidate(secretName: String) {
        guard let conn = try? db.connection else { return }
        _ = try? conn.run(cacheTable.filter(secretNameCol == secretName).delete())
    }

    static func buildMessage(timestampData: Data, callerID: String, secretName: String) -> Data {
        var message = timestampData
        let callerData = Data(callerID.utf8)
        let nameData = Data(secretName.utf8)
        var callerLen = UInt32(callerData.count).bigEndian
        message.append(Data(bytes: &callerLen, count: 4))
        message.append(callerData)
        message.append(nameData)
        return message
    }
}
